src/workspace/resolver.rs

```diff
@@ -240,18 +242,135 @@ pub enum WorkspaceResolvePkgJsonFolderErrorKind {
 }
 
 #[derive(Debug)]
-pub struct WorkspaceResolver {
+struct CompilerOptionsRootDirsResolver<TSys: FsMetadata> {
+  root_dirs_from_root: Vec<Url>,
+  root_dirs_by_member: BTreeMap<Url, Option<Vec<Url>>>,
+  sys: TSys,
+}
+
+impl<TSys: FsMetadata> CompilerOptionsRootDirsResolver<TSys> {
+  fn from_workspace(workspace: &Workspace, sys: TSys) -> Self {
+    fn get_root_dirs(
+      config_file: &ConfigFile,
+      dir_url: &Url,
+    ) -> Option<Vec<Url>> {
+      let dir_path = url_to_file_path(dir_url).ok()?;
+      let root_dirs = config_file
+        .json
+        .compiler_options
+        .as_ref()?
+        .as_object()?
+        .get("rootDirs")?
+        .as_array()?
+        .iter()
+        .filter_map(|s| {
+          url_from_directory_path(&dir_path.join(s.as_str()?)).ok()
+        })
+        .collect();
+      Some(root_dirs)
+    }
+    let root_deno_json = workspace.root_deno_json();
+    let root_dirs_from_root = root_deno_json
+      .and_then(|c| {
+        let root_dir_url = c.specifier.join(".").ok()?;
+        get_root_dirs(c, &root_dir_url)
+      })
+      .unwrap_or_default();
+    let root_dirs_by_member = workspace
+      .resolver_deno_jsons()
+      .filter_map(|c| {
+        if let Some(root_deno_json) = root_deno_json {
+          if c.specifier == root_deno_json.specifier {
+            return None;
+          }
+        }
+        let dir_url = c.specifier.join(".").ok()?;
```

Comment:
Maybe it would be beneficial if we created some diagnostics for these to display as warnings? That way someone can figure out when they've done something wrong. Would be nice to get that in this PR if it's not too much effort, but otherwise let's open an issue for a follow-up if that would be beneficial.